nixos/modules/services/web-apps/discourse.nix

```diff
@@ -0,0 +1,1079 @@
+{ config, options, lib, pkgs, utils, ... }:
+
+let
+  json = pkgs.formats.json {};
+
+  cfg = config.services.discourse;
+
+  postgresqlPackage = if config.services.postgresql.enable then
+                        config.services.postgresql.package
+                      else
+                        pkgs.postgresql;
+
+  # We only want to create a database if we're actually going to connect to it.
+  databaseActuallyCreateLocally = cfg.database.createLocally && cfg.database.host == null;
+
+  tlsEnabled = (cfg.enableACME
+                || cfg.sslCertificate != null
+                || cfg.sslCertificateKey != null);
+
+  discourseEnv = {
+    HOME = "/run/discourse/home";
+    RAILS_ENV = "production";
+    UNICORN_LISTENER = "/run/discourse/sockets/unicorn.sock";
+    UNICORN_TIMEOUT = builtins.toString cfg.unicornTimeout;
+    UNICORN_SIDEKIQS = builtins.toString cfg.sidekiqProcesses;
+  };
+
+  discourseDeps = [
+    postgresqlPackage
+
+    # For backups, themes and assets
+    cfg.package.ruby
+    pkgs.gzip
+    pkgs.gnutar
+    pkgs.git
+    pkgs.brotli
+
+    # Misc required system utils
+    pkgs.which
+    pkgs.procps       # For ps and kill
+    pkgs.util-linux   # For renice
+    pkgs.gawk
+
+    # Image optimization
+    pkgs.imagemagick
+    pkgs.optipng
+    pkgs.pngquant
+    pkgs.libjpeg
+    pkgs.jpegoptim
+    pkgs.gifsicle
+    pkgs.nodePackages.svgo
+  ];
+
+  discourse-rake = pkgs.runCommandNoCC "discourse-rake" {
+    nativeBuildInputs = [ pkgs.makeWrapper ];
+  } ''
+    mkdir -p $out/bin
+    makeWrapper ${cfg.package.rubyEnv}/bin/rake $out/bin/discourse-rake \
+        ${lib.concatStrings (lib.mapAttrsToList (name: value: "--set ${name} '${value}' ") discourseEnv)} \
+        --prefix PATH : ${lib.makeBinPath discourseDeps} \
+        --set RAKEOPT '-f ${cfg.package}/share/discourse/Rakefile' \
+        --run 'cd ${cfg.package}/share/discourse'
+  '';
+in
+{
+  options = {
+    services.discourse = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Enable the Discourse service.
+        '';
+      };
+
+      user = lib.mkOption {
+        type = lib.types.str;
+        default = "discourse";
+        description = ''
+          User account under which Discourse runs.
+        '';
+      };
+
+      group = lib.mkOption {
+        type = lib.types.str;
+        default = "discourse";
+        description = ''
+          Group account under which Discourse runs.
+        '';
+      };
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.discourse;
+        defaultText = "pkgs.discourse";
+        description = ''
+          The discourse package to use.
+        '';
+      };
+
+      hostname = lib.mkOption {
+        type = lib.types.str;
+        default = config.networking.hostName;
```

Comment:
```suggestion
        default = config.networking.fqdn;
        defaultText = "config.networking.fqdn";
```

Reply:
Yep, that's more reasonable. It required some working around `config.networking.domain` being `null`, but is definitely more what we want.